php-classes/Validators.class.php

```diff
@@ -56,10 +56,15 @@ public static function number($number, $options = array())
             && (($options['max'] === false) || ($number <= $options['max']));
     }
 
-    public static function email($email)
+    public static function email($email, $options = [])
     {
+        $options = array_merge([
+            'domain' => null
+        ], $options);
+
         return !empty($email)
-            && preg_match('/^[_a-zA-Z0-9-+]+(\.[_+a-zA-Z0-9-]+)*@[a-zA-Z0-9-]+(\.[a-zA-Z0-9-]+)*(\.[a-zA-Z]{2,})$/', $email);
+            && preg_match('/^[_a-zA-Z0-9-+]+(\.[_+a-zA-Z0-9-]+)*@[a-zA-Z0-9-]+(\.[a-zA-Z0-9-]+)*(\.[a-zA-Z]{2,})$/', $email)
+            && (empty($options['domain']) || preg_match('/^[_a-zA-Z0-9-+]+(\.[_+a-zA-Z0-9-]+)*@'.$options['domain'].'$/', $email));
```

Comment:
@nbey instead of duplicating the first part of this regex, let's dynamically build the regex string.

Also, you need to use `preg_quote` on any uncontrolled string you're injecting into a regex

So do something like
```
$pattern = '[_a-zA-Z0-9-+]+(\.[_+a-zA-Z0-9-]+)*@';

if (empty($options['domain'])) {
    $pattern .= '[a-zA-Z0-9-]+(\.[a-zA-Z0-9-]+)*(\.[a-zA-Z]{2,})';
} else {
    $pattern .= preg_quote($options['domain']);
}

return !empty($email) && preg_match('/^'.$pattern.'$/', $email);
```